libsolidity/formal/CHC.cpp

```diff
@@ -1386,11 +1386,56 @@ void CHC::defineExternalFunctionInterface(FunctionDefinition const& _function, C
 	state().addBalance(state().thisAddress(), k.currentValue());
 
 	if (encodeExternalCallsAsTrusted())
+	{
 		// If the contract has state variables that are addresses to other contracts,
 		// we need to encode the fact that those contracts may have been called in between
 		// transactions to _contract.
-		// We do that by adding nondet_interface constraints for those contracts.
-		addNondetCalls(*m_currentContract);
+		//
+		// We do that by adding nondet_interface constraints for those contracts,
+		// in the last line of this if block.
+		//
+		// If there are state variables of container types like structs or arrays
+		// that indirectly contain contract types, we havoc the state for simplicity,
+		// in the first part of this block.
+		// TODO: This could actually be supported.
+		// For structs: simply collect the SMT expressions of all the indirect contract type members.
+		// For arrays: more involved, needs to traverse the array symbolically and do the same for each contract.
+		// For mappings: way more complicated if the element type is a contract.
+		auto hasContractSubType = [&](VariableDeclaration const* _var) -> bool {
+			bool foundContract = false;
+			solidity::util::BreadthFirstSearch<Type const*> bfs{{_var->type()}};
+			bfs.run([&](auto _type, auto&& _addChild) {
+				if (dynamic_cast<ContractType const*>(_type))
```

Comment:
I'm also still a bit fuzzy about ``address``es in all of this... I'll write a separate comment about it.